### PR TITLE
[github] Get GitHub issue comments

### PR DESCRIPTION
**Pull request by valeriocos**
This pull request contains code to fetch issue comments (see #53).

### CODE REVIEW

File: perceval/backends/core/github.py [hunk 1]
```diff
@@ -85,6 +86,31 @@ def __init__(self, owner=None, repository=None,
...
+            self._push_cache_queue('{COMMENTS}')
+            self._push_cache_queue(raw_comments)
+            self._flush_cache_queue()
+            for comment in json.loads(raw_comments):
+                comment["user_data"] = self.__get_user(comment['user']['login'])
+                comments.append(comment)
+
+        return comments
+
+    def __get_issue_assignees(self, raw_assignees):
```
Comment (sduenas) [inline perceval/backends/core/github.py]: This should be removed and included in another PR. It's a new feature not related with this PR.

File: perceval/backends/core/github.py [hunk 2]
```diff
@@ -85,6 +86,31 @@ def __init__(self, owner=None, repository=None,
                                    sleep_for_rate, min_rate_to_sleep)
         self._users = {}  # internal users cache
 
+    def __get_issue_comments(self, issue_number):
+        """ Get issue comments """
+        comments = []
```
Comment (sduenas) [inline perceval/backends/core/github.py]: Leave a blank line between the docstring and the code.

File: perceval/backends/core/github.py [hunk 3]
```diff
@@ -120,83 +147,112 @@ def fetch(self, from_date=DEFAULT_DATETIME):
...
 
         cache_items = self.cache.retrieve()
-
+        current_issue = None
         issues = None
 
         while True:
             try:
                 raw_item = next(cache_items)
+
```
Comment (sduenas) [inline perceval/backends/core/github.py]: I think you should split this work in several and small methods. It would be easy to read and to understand. One method for processing users, one for assignees and so on.

File: perceval/backends/core/github.py [hunk 4]
```diff
@@ -318,17 +378,16 @@ def __send_request(self, url, params=None, headers=None):
         logger.debug("Rate limit: %s" % (self.rate_limit))
         return r
 
-    def get_issues(self, start=None):
+    def get_paginated_items(self, path, payload_type, start=None):
```
Comment (sduenas) [inline perceval/backends/core/github.py]: I like the idea of this method but in my opinion the method for fetching issues should not be removed. We can have one private method called `_fetch_items` and several methods like `issues` or `comments` that calling to `fetch_items` will download the data. For instance, see how it's implemented in `meetup` backend: https://github.com/grimoirelab/perceval/blob/master/perceval/backends/core/meetup.py#L375

Comment (sduenas) [reply perceval/backends/core/github.py]: In any case, this functionality should be included in a separated commit.

File: perceval/backends/core/github.py [hunk 5]
```diff
@@ -85,6 +86,32 @@ def __init__(self, owner=None, repository=None,
...
+            self._push_cache_queue('{COMMENTS}')
+            self._push_cache_queue(raw_comments)
+            self._flush_cache_queue()
+            for comment in json.loads(raw_comments):
+                comment["user_data"] = self.__get_user(comment['user']['login'])
+                comments.append(comment)
+
+        return comments
+
+    def __get_issue_assignee(self, raw_assignee):
```
Comment (sduenas) [inline perceval/backends/core/github.py]: As I said before, you should remove this feature.

File: perceval/backends/core/github.py [hunk 6]
```diff
@@ -120,83 +148,133 @@ def fetch(self, from_date=DEFAULT_DATETIME):
...
             for issue in issues:
-                for field in ['user', 'assignee']:
+                for field in TARGET_ISSUE_FIELDS:
+                    issue[field + "_data"] = {}
+
                     if issue[field]:
-                        issue[field + "_data"] = self.__get_user(issue[field]['login'])
-                    else:
-                        issue[field + "_data"] = {}
+                        if field == "user":
```
Comment (sduenas) [inline perceval/backends/core/github.py]: For this types of strings, please use single quotes. We usually use them. You should change them in other places of this patch.

File: perceval/backends/core/github.py [hunk 7]
```diff
@@ -120,83 +148,133 @@ def fetch(self, from_date=DEFAULT_DATETIME):
...
+                            issue[field + "_data"] = self.__get_user(issue[field]['login'])
+                        elif field == "assignee":
+                            issue[field + "_data"] = self.__get_issue_assignee(issue[field])
+                        elif field == "comments":
+                            issue[field + "_data"] = self.__get_issue_comments(issue["number"])
+                self._push_cache_queue('{}')
+                self._flush_cache_queue()
                 yield issue
 
+    def __fetch_issues_from_cache(self, cache_items):
```
Comment (sduenas) [inline perceval/backends/core/github.py]: Private methods should be after public methods.

File: perceval/backends/core/github.py [hunk 8]
```diff
@@ -120,83 +148,133 @@ def fetch(self, from_date=DEFAULT_DATETIME):
...
+        """Fetch issues from cache"""
+
+        raw_content = next(cache_items)
+        issues = (i for i in json.loads(raw_content))
```
Comment (sduenas) [inline perceval/backends/core/github.py]: Why a tuple and not a list?

File: perceval/backends/core/github.py [hunk 9]
```diff
@@ -120,83 +148,133 @@ def fetch(self, from_date=DEFAULT_DATETIME):
...
             try:
                 raw_item = next(cache_items)
+
+                if raw_item == '{ISSUES}':
+                    issues = self.__fetch_issues_from_cache(cache_items)
+
+                    if issues:
+                        current_issue = next(issues)
+                        self.__init_extra_issue_fields(current_issue)
+
```
Comment (sduenas) [inline perceval/backends/core/github.py]: Remove blank lines between `if/elif/else` statements. It improves the reading of the code.

File: perceval/backends/core/github.py [hunk 10]
```diff
@@ -120,83 +148,133 @@ def fetch(self, from_date=DEFAULT_DATETIME):
...
+                elif raw_item == '{}':
+                    yield current_issue
+                    current_issue = next(issues)
+                    self.__init_extra_issue_fields(current_issue)
+
             except StopIteration:
                 if issues:
-                    for issue in self.__build_issues(issues):
-                        yield issue
+                    (i for i in issues)
```
Comment (sduenas) [inline perceval/backends/core/github.py]: What is the rationale behind this?

File: perceval/backends/core/github.py (hunk 5 above)
Comment (valeriocos) [reply perceval/backends/core/github.py]: the code in the master in branch already dealt with issue assignee. I removed the feature issue_assignees

File: perceval/backends/core/github.py (hunk 8 above)
Comment (valeriocos) [reply perceval/backends/core/github.py]:
```python
(i for i in json.loads(raw_content))
``` 
should be the pythonic way to return a generator

File: perceval/backends/core/github.py (hunk 10 above)
Comment (valeriocos) [reply perceval/backends/core/github.py]: My mistake, this part can be replaced by 
````python
except StopIteration:
    break
````

File: perceval/backends/core/github.py (hunk 5 above)
Comment (sduenas) [reply perceval/backends/core/github.py]: Ok! I forgot about that.

File: perceval/backends/core/github.py [hunk 11]
```diff
@@ -120,83 +102,174 @@ def fetch(self, from_date=DEFAULT_DATETIME):
...
+        """ Get issue assignee """
+
+        self._push_cache_queue('{ASSIGNEE}')
+        self._push_cache_queue(raw_assignee)
+        self._flush_cache_queue()
+        assignee = self.__get_user(raw_assignee['login'])
+
+        return assignee
+
+    def __get_user(self, login):
```
Comment (sduenas) [inline perceval/backends/core/github.py]: Has this method change? There is a cache in memory to avoid many trips to the server when requesting for users. You can find the dict in `self._users`. This method should check that dict first.

Comment (valeriocos) [reply perceval/backends/core/github.py]: No it hasn't. This method calls
 ```python 
self.client.get_user(login)
```
 which takes care of caching the users

Comment (sduenas) [reply perceval/backends/core/github.py]: Ok!

**Comment by valeriocos** [comment]
:tada: :tada: :tada:
